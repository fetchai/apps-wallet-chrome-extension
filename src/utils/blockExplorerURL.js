import { MAINNET_BLOCKEXPLORER, NETWORKS_ENUM, STORAGE_ENUM, TESTNET_BLOCKEXPLORER } from '../constants'


/**
 * return block explorer url (with accounts path) based on whichever network is users selected network.
 *
 * @returns {string}
 */


const blockExplorerURL = (url_bit = 'accounts/') => {
  const network = localStorage.getItem(STORAGE_ENUM.SELECTED_NETWORK)
  if (network === NETWORKS_ENUM.TESTNET) return TESTNET_BLOCKEXPLORER + url_bit
  else if (network === NETWORKS_ENUM.MAINNET) return MAINNET_BLOCKEXPLORER + url_bit
  else if (network === NETWORKS_ENUM.LOCALHOST) return ''
  else throw new Error(`the network from storage : ${network} is not an expected network`)
}

export { blockExplorerURL }