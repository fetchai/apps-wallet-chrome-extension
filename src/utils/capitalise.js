const capitalise = s => {
  if (typeof s !== 'string' || s.length === 0) return ''
  return s.charAt(0).toUpperCase() + s.slice(1)
}

export { capitalise }
