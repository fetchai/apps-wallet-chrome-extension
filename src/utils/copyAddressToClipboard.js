
async function copyToClipboard (str) {
  return new Promise((resolve, reject) =>
  {
    navigator.clipboard.writeText(str).then(() => {
      resolve(true)
    }, () => {
      reject(false)
    })
  })
  }


export {copyToClipboard}