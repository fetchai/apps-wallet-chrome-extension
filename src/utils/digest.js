import { createHash } from 'crypto'

const digest = (address_raw) => {
  const hash_func = createHash('sha256')
  hash_func.update(address_raw)
  return hash_func.digest()
}

export { digest }
