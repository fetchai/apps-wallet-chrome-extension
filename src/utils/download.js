import Storage from '../services/storage'
import { KEY_FILE_NAME, STORAGE_ENUM } from '../constants'

/**
   * Causes download of encrypted key file as json file, taking key file from storage.
   *
   * @returns {Promise<void>}
   */

  const download = async () => {
    const json_str = await Storage.getItem(STORAGE_ENUM.KEY_FILE)
    const element = document.createElement('a')
    const file = new Blob([json_str], { type: 'text/plain' })
    element.href = URL.createObjectURL(file)
    element.download = KEY_FILE_NAME
    document.body.appendChild(element)
    element.click()
  }

  export { download }