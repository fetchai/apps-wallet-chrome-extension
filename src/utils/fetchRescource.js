/*global chrome*/
import { EXTENSION } from '../constants'

/**
 * Extensions cannot make CORs requests in content scripts therefore such requests are made in background script and results passed to content script.
 * This has same API as HTML5 Fetch Api https://www.chromium.org/Home/chromium-security/extension-content-script-fetches so we can easily switch to run in browser.
 *
 * @param input
 * @returns {Promise<unknown>}
 */

function proxyToBackground (input, init) {
  return new Promise((resolve, reject) => {
    chrome.runtime.sendMessage({ input, init }, messageResponse => {
      const [response, error] = messageResponse
      if (response === null) {
        reject(null)
      } else {
        // Use undefined on a 204 - No Content
        const body = response.body ? new Blob([response.body]) : undefined
        resolve(new Response(body, {
          status: response.status,
          statusText: response.statusText,
        }))
      }
    })
  })
}

/**
 * This is the switch that allows us to run in browser making requests, or background script we pass them to background.
 *
 * @param input
 * @param init
 * @returns {*}
 */
function fetchResource (input, init) {
  return EXTENSION ? proxyToBackground(input, init) : fetch(input, init)
}

export { fetchResource }
