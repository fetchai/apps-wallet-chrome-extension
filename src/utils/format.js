   const format = (val) => val.substring(0, 12) + "..."

export {format}