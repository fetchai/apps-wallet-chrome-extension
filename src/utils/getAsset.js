/*global chrome*/
import {EXTENSION} from "../constants";

/**
 * Different is returned URI for asset stored in assets folder
 * depending on if we run as an extension or in browser. If this throws an error on compilation
 * check whether your EXTENSION flag is correct for the environment you are running in.
 *
 * @param name
 * @returns {string}
 */
const getAssetURI = name => EXTENSION ? chrome.runtime.getURL("/assets/" + name) :  `./assets/${name}`

export {getAssetURI}