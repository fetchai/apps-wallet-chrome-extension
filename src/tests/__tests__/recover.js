import React from 'react'
import ReactDOM from 'react-dom'
import chrome from 'sinon-chrome'
import { cleanup, fireEvent, render, wait } from '@testing-library/react'
import '@testing-library/jest-dom'
import Router, { clear } from '../../services/router'
import {
  click_button,
  form_write_value,
  STRONG_PASSWORD,
  untilPromises,
  WEAK_PASSWORD
} from './../utils/helper'

import Recover from '../../views/recover'
import {
  ENCRYPTED_KEY_FILE,
  EXAMPLE_ADDRESS,
  EXAMPLE_INVALID_ADDRESS,
  INVALID_ENCRYPTED_KEY_FILE
} from '../utils/helper'
import { mockAccountPageConstructor } from '../utils/mockAccountPageConstructor'

const FILE_REQUIRED_ERROR_MESSAGE = "File required";
const INCORRECT_PASSWORD_OR_ADDRESS_ERROR_MESSAGE = 'Incorrect Password or Address';
const UNDECRYPTABLE_ERROR_MESSAGE = 'Unable to decrypt';
const INVALID_ADDRESS_ERROR_MESSAGE = 'Invalid address';
const INCORRECT_FILE_TYPE_ERROR_MESSAGE = 'Incorrect file type';
const PASSWORD_REQUIRED_ERROR_MESSAGE = 'Password required';
const FILE_NAME = 'chucknorris.png'
const NON_JSON_FILE_CONTENTS = '(⌐□_□)';
const UNREADABLE_FILE_ERROR_MESSAGE = 'Unable to read file';
const WEAK_PASSWORD_ERROR_MESSAGE = 'Incorrect Password: Password too weak (14 chars including letter, number, special character and uppercase letter';

const add_file = (getByTestId, contents, name) => {
const file = new File([contents], name, { type: 'image/png', })
  const inputEl = getByTestId('file_input')
  //https://github.com/kentcdodds/react-testing-library-examples/pull/1/files
  Object.defineProperty(inputEl, 'files', {
    value: [file],
  })

  fireEvent.change(inputEl)
}


const submit_form = async ({getByTestId, file_name = null, file_contents = null, password = null, address = null }) => {

  if(file_name !== null || file_contents !== null) {
    await add_file(getByTestId, file_contents || file_name)
  }

  if(password !== null){
    await form_write_value(getByTestId, 'password_input', password)
  }

  if(address !== null){
      await form_write_value(getByTestId, 'address_input', EXAMPLE_INVALID_ADDRESS)
  }
 await untilPromises()
 await click_button(getByTestId, 'upload_button')
}

describe(':Recover', () => {

  beforeAll(() => {
    global.chrome = chrome
  })

  beforeEach(() => {
     jest.resetAllMocks();
     jest.clearAllMocks()
  })

  afterEach(() => {
    cleanup()
    clear()
  })

  test.skip('Create renders without crashing', () => {
    const div = document.createElement('div')
    ReactDOM.render(<Recover/>, div)
    ReactDOM.unmountComponentAtNode(div)
  })

  test.skip('test initially does not show warning div', () => {
    const { getByTestId, } = render(<Router><Recover/></Router>)
    expect(getByTestId("warning_container")).not.toBeVisible()
  })

  test.skip('test back button redirects to terms', async () => {
    const { getByTestId, getAllByTestId } = render(<Router><Recover/></Router>)
    await click_button(getByTestId, 'back_button')
    // tests the account component is now mounted by router.
    const app = getAllByTestId('terms')
    expect(app.length).toBe(1)
  })

  test.skip('test empty upload form shows file required error message', async () => {
    const { getByTestId, } = render(<Recover/>)
    await click_button(getByTestId, 'upload_button')
    // tests the account component is now mounted by router.
      await untilPromises()
    expect(getByTestId('output_error')).toHaveTextContent(FILE_REQUIRED_ERROR_MESSAGE)
  })


  test.skip('test non json file type shows incorrect file type error message', async () => {
    const { getByTestId } = render(<Router><Recover/></Router>)
    await submit_form({getByTestId: getByTestId, file_name: FILE_NAME, file_contents: NON_JSON_FILE_CONTENTS, password: STRONG_PASSWORD})
    await untilPromises()
   // wait for appearance
    await wait(() => {
   expect(getByTestId('output_error')).toHaveTextContent(INCORRECT_FILE_TYPE_ERROR_MESSAGE)
  })
  })

  test('key file that cannot be decrypted by javascript api in otherwise valid form shows correct error message', async () => {
    const { getByTestId } = render(<Router><Recover/></Router>)
    await submit_form({getByTestId: getByTestId, file_name: FILE_NAME, file_contents: INVALID_ENCRYPTED_KEY_FILE, password: STRONG_PASSWORD})
    await untilPromises()
    await wait(() => {
   expect(getByTestId('output_error')).toHaveTextContent(UNDECRYPTABLE_ERROR_MESSAGE)
  })
  })

  test.skip('test invalid address shows invalid address error message', async () => {
    //todo reimplement valid address test
    const { getByTestId  } = render(<Router><Recover/></Router>)
    await submit_form({getByTestId: getByTestId, file_name: FILE_NAME, file_contents: ENCRYPTED_KEY_FILE, password: STRONG_PASSWORD, address: EXAMPLE_INVALID_ADDRESS})
    await untilPromises()
    await wait(() => {
   expect(getByTestId('output_error')).toHaveTextContent(INVALID_ADDRESS_ERROR_MESSAGE)
  })
  })

  test.skip('test file reader onerror shows correct error message', async () => {
    const { getByTestId, } = render(<Router><Recover/></Router>)
   const spy = jest.spyOn(FileReader.prototype, 'readAsText').mockImplementation(function(){
       this.onerror()
      })

     await submit_form({getByTestId: getByTestId, file_name: FILE_NAME, file_contents: ENCRYPTED_KEY_FILE, password: STRONG_PASSWORD, address: EXAMPLE_INVALID_ADDRESS})
    spy.mockClear()
      // wait for appearance
  await wait(() => {
   expect(getByTestId('output_error')).toHaveTextContent(UNREADABLE_FILE_ERROR_MESSAGE)
  })

spy.mockRestore();
  })

  test.skip('test file decrypts to wrong address shows correct error message', async () => {
    const { getByTestId } = render(<Router><Recover/></Router>)
      await submit_form({getByTestId: getByTestId, file_name: FILE_NAME, file_contents: ENCRYPTED_KEY_FILE, password: STRONG_PASSWORD, address: EXAMPLE_ADDRESS})
   await untilPromises()
      // wait for appearance
  await wait(() => {
   expect(getByTestId('output_error')).toHaveTextContent(INCORRECT_PASSWORD_OR_ADDRESS_ERROR_MESSAGE)
  })
  })

  test.skip('test weak password shows weak password error message', async () => {
 const { getByTestId } = render(<Router><Recover/></Router>)
   await submit_form({getByTestId: getByTestId, file_name: FILE_NAME, file_contents: ENCRYPTED_KEY_FILE, password: WEAK_PASSWORD, address: EXAMPLE_ADDRESS})
   await untilPromises()
      // wait for appearance
  await wait(() => {
   expect(getByTestId('output_error')).toHaveTextContent(INCORRECT_PASSWORD_OR_ADDRESS_ERROR_MESSAGE)
  })
  })

  test.skip('test submitting otherwise valid file and strong password but without providing your address shows you warning', async () => {
    const { getByTestId } = render(<Router><Recover/></Router>)
    await submit_form({getByTestId: getByTestId, file_name: FILE_NAME, file_contents: ENCRYPTED_KEY_FILE, password: STRONG_PASSWORD})
    await untilPromises()
      // wait for appearance
     await wait(async () => {
    // the warning container with warning message is shown
      expect(getByTestId("warning_container")).toBeVisible()
  })
  })

  test.skip('test no password in otherwise valid form shows password required error message', async () => {
    const { getByTestId } = render(<Router><Recover/></Router>)
    await submit_form({getByTestId: getByTestId, file_name: FILE_NAME, file_contents: ENCRYPTED_KEY_FILE, password: ""})
    await untilPromises()
    expect(getByTestId('output_error')).toHaveTextContent(PASSWORD_REQUIRED_ERROR_MESSAGE)
  })

  test.skip('test weak password in otherwise valid form shows weak password error message', async () => {
    const { getByTestId } = render(<Router><Recover/></Router>)
    await submit_form({getByTestId: getByTestId, file_name: FILE_NAME, file_contents: ENCRYPTED_KEY_FILE, password: WEAK_PASSWORD})
    await untilPromises()
     expect(getByTestId('output_error')).toHaveTextContent(WEAK_PASSWORD_ERROR_MESSAGE)
  })


  test.skip('test submitting otherwise valid file and strong password but without providing your address shows you warning', async () => {
    mockAccountPageConstructor()
    const { getByTestId, getAllByTestId } = render(<Router><Recover/></Router>)
    await submit_form({getByTestId: getByTestId, file_name: FILE_NAME, file_contents: ENCRYPTED_KEY_FILE, password: STRONG_PASSWORD})
    await untilPromises()
    await click_button(getByTestId, 'next_button')
    await untilPromises()
    await wait(async () => {
     // tests the account component is now mounted by router.
    const app = getAllByTestId('account')
    expect(app.length).toBe(1)
  })
  })


  test.skip('test back button on warning message shows you form again', async () => {
   const { getByTestId, getAllByTestId } = render(<Router><Recover/></Router>)
    await submit_form({getByTestId: getByTestId, file_name: FILE_NAME, file_contents: ENCRYPTED_KEY_FILE, password: STRONG_PASSWORD})
    await untilPromises()
    await click_button(getByTestId, 'second_back_button')
    await untilPromises()
    await wait(async () => {
     // tests the account component is now mounted by router.
    const app = getAllByTestId('recover')
    expect(app.length).toBe(1)
  })  })

  test.skip('test file decrypts to correct address logs in and redirects to accounts page', async () => {
    mockAccountPageConstructor()
    const { getByTestId, getAllByTestId } = render(<Router><Recover/></Router>)
    const CORRECT_ADDRESS = "AXPMC2wMWr3tjDBuSXi4FEsn8YU2ysMASZMpk7qTEBthWwSwn";

   add_file(getByTestId, ENCRYPTED_KEY_FILE, FILE_NAME)
   await form_write_value(getByTestId, 'password_input', STRONG_PASSWORD)
   await form_write_value(getByTestId, 'address_input', CORRECT_ADDRESS)
   await click_button(getByTestId, 'upload_button')
   await untilPromises()
    // tests the account component is now mounted by router.
   await wait(() => {
      const app = getAllByTestId('account')
      expect(app.length).toBe(1)
    })
  })


})