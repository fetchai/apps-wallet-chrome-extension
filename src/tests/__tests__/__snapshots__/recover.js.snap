// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`:Recover initial snapshot test 1`] = `
<div
  className="OverlayMain"
  data-testid="recover"
  id="my-extension-root-inner"
>
  <div
    className="OverlayMainInner"
  >
    <h2>
      Recover
    </h2>
    <hr />
    <div
      className=""
      style={
        Object {
          "height": "auto",
          "opacity": 1,
          "overflow": "unset",
          "transition": "height 500ms ease-in-out,opacity 500ms ease-in-out,background 500ms ease-in-out",
        }
      }
    >
      <form
        className="recover-form"
        id="form"
      >
        <legend
          className="recover-legend"
        >
          Upload File with Password
        </legend>
        <button
          className="recover-input   upload-button"
          onClick={[Function]}
        >
          
        </button>
        <input
          className="hide"
          id="file"
          label="file"
          onChange={[Function]}
          type="file"
        />
        <input
          className="recover-input "
          id="password"
          name="password"
          onChange={[Function]}
          placeholder="Password"
          type="password"
          value=""
        />
        <input
          className="recover-input recover-address "
          id="address"
          label="address"
          name="address"
          onChange={[Function]}
          placeholder="Address (optional)"
          type="text"
          value=""
        />
        <output
          className="recover-output "
          id="output"
          type="text"
        >
          
        </output>
        <div
          className="small-button-container"
        >
          <button
            className="recover-back-button"
            onClick={[Function]}
            type="button"
          >
            Back
          </button>
          <button
            className="recover-upload-button"
            onClick={[Function]}
            type="submit"
          >
            Upload
          </button>
        </div>
      </form>
    </div>
    <div
      className=""
      style={
        Object {
          "height": 0,
          "opacity": 0,
          "overflow": "hidden",
          "transition": "height 500ms ease-in-out,opacity 500ms ease-in-out,background 500ms ease-in-out",
        }
      }
    >
      <div
        className="recover-warning-container"
      >
        <p>
          Warning! Not providing your public address means that recovery won't error if your password is incorrect. 
        </p>
        <p>
          It will instead decrypt to a random incorrect address (probably without funds).
        </p>
        <p>
          If this happens don't panic but clear your local storage and try again.
        </p>
        <p>
          Click "Next" to proceed anyway or "Back" to return to the upload form.
        </p>
      </div>
      <div
        className="small-button-container"
      >
        <button
          className="recover-back-button"
          onClick={[Function]}
          type="button"
        >
          Back
        </button>
        <button
          className="recover-upload-button"
          onClick={[Function]}
          type="submit"
        >
          Next
        </button>
      </div>
    </div>
  </div>
</div>
`;
